src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.DelegateTemplateSymbol.cs

```diff
@@ -0,0 +1,174 @@
+                    }
+
+                    // if we are given Void type the method returns Void, otherwise its return type is the last type parameter of the delegate:
+                    var returnType = TypeWithAnnotations.Create(voidReturnTypeOpt ?? typeParams[parameterCount]);
+                    var returnRefKind = (refKinds.IsNull || voidReturnTypeOpt is { }) ? RefKind.None : refKinds[parameterCount];
+
+                    var method = new SynthesizedDelegateInvokeMethod(containingType, parameterTypes, parameterRefKinds, returnType, returnRefKind);
+                    parameterRefKinds.Free();
+                    parameterTypes.Free();
+                    return method;
+                }
+
+                static ImmutableArray<TypeParameterSymbol> createTypeParameters(AnonymousDelegateTemplateSymbol containingType, int parameterCount, bool returnsVoid)
+                {
+                    var typeParameters = ArrayBuilder<TypeParameterSymbol>.GetInstance(parameterCount + (returnsVoid ? 0 : 1));
+                    for (int i = 0; i < parameterCount; i++)
+                    {
+                        typeParameters.Add(new AnonymousTypeManager.AnonymousTypeParameterSymbol(containingType, i, "T" + (i + 1)));
+                    }
+
+                    if (!returnsVoid)
+                    {
+                        typeParameters.Add(new AnonymousTypeManager.AnonymousTypeParameterSymbol(containingType, parameterCount, "TResult"));
+                    }
+
+                    return typeParameters.ToImmutableAndFree();
+                }
+            }
+
+            internal AnonymousDelegateTemplateSymbol(AnonymousTypeManager manager, AnonymousTypeDescriptor typeDescr, ImmutableArray<TypeParameterSymbol> typeParametersToSubstitute)
+                : base(manager, typeDescr.Location)
+            {
+                // AnonymousTypeOrDelegateComparer requires an actual location.
+                Debug.Assert(SmallestLocation != null);
+                Debug.Assert(SmallestLocation != Location.None);
+
+                IsParameterizedDelegateType = false;
+
+                TypeMap typeMap;
+                int typeParameterCount = typeParametersToSubstitute.Length;
+                if (typeParameterCount == 0)
+                {
+                    TypeParameters = ImmutableArray<TypeParameterSymbol>.Empty;
+                    typeMap = TypeMap.Empty;
+                }
+                else
+                {
+                    var typeParameters = ArrayBuilder<TypeParameterSymbol>.GetInstance(typeParameterCount);
+                    for (int i = 0; i < typeParameterCount; i++)
+                    {
+                        typeParameters.Add(new AnonymousTypeParameterSymbol(this, i, "T" + (i + 1)));
+                    }
+                    TypeParameters = typeParameters.ToImmutableAndFree();
+                    typeMap = new TypeMap(typeParametersToSubstitute, TypeParameters, allowAlpha: true);
```

Comment:
>allowAlpha: true [](http://example.com/codeflow?start=86&length=16)

Why is this important?

Reply:
`ClosureConversion.RewriteLambdaOrLocalFunction()` rewrites nested functions including the types of any locals in the functions. If the local type is an `AnonymousDelegatePublicSymbol`, that will substitute the types of the associated `TypeDescriptor` with `SynthesizedSubstitutedTypeParameterSymbol` from the rewritten nested function. See `SynthesizedDelegateTypes_30()` for an example.

Reply:
We are mapping definitions to definitions. Why do we need to '''allowAlpha``` here.